File: python_version/application.py
# application.py - COMP3260A2
# The main thread of the des program
#
# Authors: Jay Rovacsek, Cody Lewis
# Since: 12-MAY-2018
import sys
import traceback
import os
import des

def parse_file(file): # parse the input file
    values = {}
    try:
        with open(file) as f:
            for line in f:
                if 'T' not in values:
                    values['T'] = line.replace("\n","")
                elif 'P' not in values:
                    values['P'] = line.replace("\n","")
                elif 'K' not in values:
                    values['K'] = line.replace("\n","")
                else:
                    break
    except FileNotFoundError:
        print("You may have spelt the filename incorrectly, or the file doesn't exist.\nPlease try again")
        return None
    finally:
        return values

if __name__ == "__main__": # IO
    if len(sys.argv) is 3:
        args = sys.argv
        del args[0]
        values = parse_file(args[0]) # takes file input (2nd argument)
        if len(values) is not 0:
            if values['T'] == '0':
                _des = des.des(values['K'],"encrypt")
                print("Encrypting using:\nPlaintext P: {}\nKey K: {}".format(values['P'], values['K']))
                text, key = _des.encrypt(values['P'])
                print("Ciphertext C: {}".format(text))
                avalanche = des.avalanche(values['P'], values['K'])
                print("Avalanche:\n{}".format(avalanche))
                try: # file output
                    with open(args[1], 'w') as f:
                        f.write("ENCRYPTION\nPlaintext P: {}\nKey K: {}\nCiphertext C: {}\nAvalanche:\n{}".format(values['P'], key, text, avalanche))
                    print("The results were saved to: {}".format(args[1]))
                except Exception:
                    print("An error occurred: {}".format(traceback.format_exc()))

            elif values['T'] == '1':
                _des = des.des(values['K'],"decrypt")
                print("Decrypting using:\nCiphertext C: {}\nKey K: {}".format(values['P'], values['K']))
                text, key = _des.decrypt(values['P'])
                print("Plaintext P: {}".format(text))
                try:
                    with open(args[1], 'w') as f: # file output
                        f.write("DECRYPTION\nCiphertext C: {}\nKey K: {}\nPlaintext P: {}".format(values['P'], key, text))
                    print("The results were saved to: {}".format(args[1]))
                except Exception:
                    print("An error occurred: {}".format(traceback.format_exc()))
            else:
                print("An expected encrypt/decrypt flag was not found, please refer to the README.md\n" +
                "For encryption use put '0' at the top of the file " +
                "\nFor decryption please put '1' at the top of the file")
        else:
            print("You may have accidentally tried to use a file that did not have a key/text pair")
    else:
        print("Length of arguments was not sufficient, please refer to the README.txt")

File: SubmissionFiles/COMP3260A2c3146220c3283349/des_system/src/bin/application.rs
use std::env; // for command line arguments
use std::fs::File; // for file opening
use std::io::prelude::*; // for file io
mod des; // The des module, located at ./des/mod.rs
use des::des::decrypt;
use des::des::encrypt;
use des::des::avalanche;
/**
 * application.rs - COMP3260A2
 * The main thread of the des program
 *
 * Authors: Jay Rovacsek, Cody Lewis
 * Since: 14-MAY-2018
 */
fn main() {
    // The main io thread
    let args: Vec<String> = env::args().collect();
    if args.len() >= 3 {
        // can only run with file and operation flags
        let filename = args[1].clone();
        let out_file = args[2].clone();
        let values = parse_file(filename); // get a tuple of the file text and key
        if values.0 == "0" {
            // Encryption methods
            println!(
                "Encrypting using:\nPlaintext P: {}\nKey K: {}",
                values.1, values.2
            );
            let cipher = encrypt(values.1.clone(), values.2.clone());
            println!("Ciphertext C: {}\nNow working on Avalanche...", cipher);
            let result = (cipher, avalanche(values.1.clone(), values.2.clone()));
            println!("Avalanche:\n{}\nFile written to: {}", result.1, out_file);
            write_results(
                out_file,
                values.1,
                values.2,
                result.0,
                values.0.clone(),
                result.1,
            ).expect("Failed to write file");
        } else if values.0 == "1" {
            // Decryption methods
            println!(
                "Decrypting using:\nCiphertext C: {}\nKey K: {}",
                values.1, values.2
            );
            let result = decrypt(values.1.clone(), values.2.clone());
            println!("Plaintext P: {}\nFile written to: {}", result, out_file);
            write_results(
                out_file,
                values.1,
                values.2,
                result,
                values.0.clone(),
                String::new(),
            ).expect("Failed to write file");
        } else {
            println!("The file you input was not in the right format, please refer to the README.txt file at the root of  the project");
        }
    } else {
        println!("Not enough arguments specified, please refer to the README.txt file at the root of the project");
    }
}

pub fn write_results(
    filename: String,
    plaintext: String,
    key: String,
    ciphertext: String,
    mode: String,
    avalanche: String,
) -> std::io::Result<()> {
    let mut f = File::create(filename)?;
    if mode == "0" {
        write!(
            f,
            "ENCRYPTION\nPlaintext P: {}\nKey K: {}\nCiphertext C: {}\nAvalanche:\n{}",
            plaintext, key, ciphertext, avalanche
        )?;
        f.sync_data()?;
    } else {
        write!(
            f,
            "DECRYPTION\nCiphertext C: {}\nKey K: {}\nPlaintext C: {}",
            plaintext, key, ciphertext
        )?;
        f.sync_data()?;
    }
    Ok(())
}
/// Input file parsing method
pub fn parse_file(filename: String) -> (String, String, String) {
    let mut f = File::open(filename).expect("Failed to read file");
    let mut contents = String::new();
    f.read_to_string(&mut contents)
        .expect("Something went wrong when reading the file");
    let (mode, text, key) = {
        let bytes = contents.as_bytes();
        let mut j = 0;
        let mut k = 0;
        let mut l = 0;
        let mut first = true;
        for (i, &item) in bytes.iter().enumerate() {
            if item == b'\n' {
                if first {
                    j = i;
                    first = false;
                } else if k == 0 {
                    k = i;
                } else {
                    l = i;
                    break;
                }
            }
        }
        (
            contents[..j].to_string(),
            contents[j + 1..k].to_string(),
            contents[k + 1..l].to_string(),
        )
    };
    (String::from(mode), String::from(text), String::from(key))
}
